File: lua/binary_tree_traverse.lua
local Node = {}
Node.__index = Node

function Node:new(data)
    local node = {
        data = data,
        left = nil,
        right = nil
    }
    setmetatable(node, Node)
    return node
end

function build_tree(depth, counter)
    if depth <= 0 then
        return nil
    end
    
    local node = Node:new(counter[1])
    counter[1] = counter[1] + 1
    node.left = build_tree(depth - 1, counter)
    node.right = build_tree(depth - 1, counter)
    return node
end

function traverse_in_order(node)
    if node == nil then
        return 0
    end
    
    local sum = 0
    sum = sum + traverse_in_order(node.left)
    sum = sum + node.data
    sum = sum + traverse_in_order(node.right)
    return sum
end

if #arg < 1 then
    print('Usage: binary_tree_traverse <depth>')
    os.exit(1)
end

local depth = tonumber(arg[1])
if not depth then
    print('Invalid number: ' .. arg[1])
    os.exit(1)
end

local counter = {0}

local start = os.clock()
local root = build_tree(depth, counter)
local sum = traverse_in_order(root)
local finish = os.clock()

local time_ms = math.floor((finish - start) * 1000)

print('Lua: binary_tree_traverse(' .. depth .. ') = ' .. sum)
print('Time: ' .. time_ms .. 'ms')

File: go/binary_tree_traverse.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

type Node struct {
	data  int
	left  *Node
	right *Node
}

func buildTree(depth int, counter *int) *Node {
	if depth <= 0 {
		return nil
	}
	
	node := &Node{data: *counter}
	*counter++
	node.left = buildTree(depth-1, counter)
	node.right = buildTree(depth-1, counter)
	return node
}

func traverseInOrder(node *Node) int {
	if node == nil {
		return 0
	}
	
	sum := 0
	sum += traverseInOrder(node.left)
	sum += node.data
	sum += traverseInOrder(node.right)
	return sum
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: binary_tree_traverse <depth>")
		return
	}

	depth, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println("Invalid number:", os.Args[1])
		return
	}

	counter := 0

	start := time.Now()
	root := buildTree(depth, &counter)
	sum := traverseInOrder(root)
	elapsed := time.Since(start)

	timeMs := elapsed.Milliseconds()

	fmt.Printf("Go: binary_tree_traverse(%d) = %d\n", depth, sum)
	fmt.Printf("Time: %dms\n", timeMs)
}